services/src/main/java/org/keycloak/services/resources/admin/AdminEventBuilder.java

```diff
@@ -0,0 +1,315 @@
+        sb.append(getResourcePath(model));
+        sb.append(resourcePath.substring(resourcePath.lastIndexOf('/')));
+        adminEvent.setResourcePath(sb.toString());
+        return this;
+    }
+    
+    public AdminEventBuilder resourcePath(Object model, String resourcePath, boolean segment) {
+        StringBuilder sb = new StringBuilder();
+        sb.append(getResourcePath(model));
+        int index = resourcePath.lastIndexOf('/');
+        int subIndex = resourcePath.lastIndexOf('/', index - 1);
+        sb.append(resourcePath.substring(subIndex));
+        adminEvent.setResourcePath(sb.toString());
+        return this;
+    }
+    
+    public AdminEventBuilder resourcePath(Object model, Object subModel, String resourcePath) {
+        StringBuilder sb = new StringBuilder();
+        sb.append(getResourcePath(model));
+        int index = resourcePath.lastIndexOf('/');
+        int subIndex = resourcePath.lastIndexOf('/', index - 1);
+        sb.append(resourcePath.substring(subIndex, index+1));
+        sb.append(getResourcePath(subModel));
+        adminEvent.setResourcePath(sb.toString());
+        return this;
+    }
+    
+
+    public void error(String error) {
+        adminEvent.setOperationType(OperationType.valueOf(adminEvent.getOperationType().name() + "_ERROR"));
+        adminEvent.setError(error);
+        send();
+    }
+    
+    public AdminEventBuilder representation(Object value) {
+        if (value == null || value.equals("")) {
+            return this;
+        }
+        try {
+            adminEvent.setRepresentation(JsonSerialization.writeValueAsString(value));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return this;
+    }
+    
+    public AdminEvent getEvent() {
+        return adminEvent;
+    }
+
+    public void success() {
+        send();
+    }
+
+    private void send() {
+        boolean includeRepresentation = false;
+        if(realm.isAdminEventsDetailsEnabled()) {
+            includeRepresentation = true;
+        }
+        adminEvent.setTime(Time.toMillis(Time.currentTime()));
+
+        if (store != null) {
+            try {
+                store.onEvent(adminEvent, includeRepresentation);
+            } catch (Throwable t) {
+                log.error("Failed to save event", t);
+            }
+        }
+        
+        if (listeners != null) {
+            for (EventListenerProvider l : listeners) {
+                try {
+                    l.onEvent(adminEvent, includeRepresentation);
+                } catch (Throwable t) {
+                    log.error("Failed to send type to " + l, t);
+                }
+            }
+        }
+    }
+    
+    private String getResourcePath(Object model) {
+
+        StringBuilder sb = new StringBuilder();
+
+        if (model instanceof RealmModel) {
+            RealmModel realm = (RealmModel) model;
+            sb.append("realms/" + realm.getId());
```

Comment:
Resource path is kept empty for realm update and delete. But, we still need "model instanceof RealmModel" because we will need realms/realm-id for new realm creation.